File: apps/host/src/lib/audit.ts
// Frontend wrapper for the audit log.
//
// Two surfaces: read (for the X-ray drawer) and write (for events that
// originate in the React side — composition assembled, action decided).
// Rust-side writes (tool calls, bus events, notifications, config
// changes) land in the same store automatically; this module is the
// frontend's voice into it.

import { invoke } from "@tauri-apps/api/core";

export interface AuditEvent {
  id: number;
  ts_ms: number;
  parent_id: number | null;
  kind: string;
  payload: unknown;
}

export interface QueryArgs {
  limit?: number;
  since_id?: number;
  kind_prefix?: string;
}

export async function queryAudit(args: QueryArgs = {}): Promise<AuditEvent[]> {
  // Tauri 2 matches JS arg names against Rust parameter names directly;
  // we use the snake_case form here to match the command signature.
  return invoke("audit_query", {
    limit: args.limit ?? null,
    since_id: args.since_id ?? null,
    kind_prefix: args.kind_prefix ?? null,
  });
}

export async function recordAudit(
  kind: string,
  payload?: unknown,
  parent_id?: number | null,
): Promise<number | null> {
  try {
    return await invoke<number>("audit_record", {
      kind,
      payload: payload ?? null,
      parent_id: parent_id ?? null,
    });
  } catch (e) {
    console.warn(`[audit] record(${kind}) failed`, e);
    return null;
  }
}

File: apps/host/src-tauri/src/commands/audit.rs
// Tauri commands for the audit log.
//
// `audit_query` is what the X-ray drawer calls to load the recent N
// events; `audit_record` lets the frontend push events into the same
// store (composition_assembled, action_decided, etc.) so the drawer is
// the single place to inspect the surface's history.

use serde_json::Value;
use tauri::State;

use crate::audit::{AuditEvent, NewEvent};
use crate::AppState;

#[tauri::command]
pub fn audit_query(
    state: State<'_, AppState>,
    limit: Option<u32>,
    since_id: Option<i64>,
    kind_prefix: Option<String>,
) -> Result<Vec<AuditEvent>, String> {
    state
        .audit
        .query_recent(
            limit.unwrap_or(200),
            since_id,
            kind_prefix.as_deref(),
        )
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn audit_record(
    state: State<'_, AppState>,
    kind: String,
    payload: Option<Value>,
    parent_id: Option<i64>,
) -> Result<i64, String> {
    let id = state.audit.record(NewEvent {
        kind,
        payload: payload.unwrap_or(Value::Null),
        parent_id,
    });
    id.ok_or_else(|| "audit write failed".to_string())
}
